src/testing/mock.rs

```diff
@@ -280,120 +170,90 @@ impl UnreliableRouter {
-                    }
-                    Poll::Pending => {
-                        break;
+        while !self.peers.is_empty() {
+            for (peer_id, peer) in self.peers.iter_mut() {
+                loop {
+                    match peer.rx.try_next() {
+                        Ok(Some(msg)) => buffer.push((*peer_id, msg)),
+                        Ok(None) => {
+                            disconnected_peers.push(*peer_id);
+                            break;
+                        }
+                        Err(_) => break,
                     }
                 }
             }
-        }
-        for peer_id in disconnected_peers {
-            self.peers.lock().remove(&peer_id);
-        }
-        for (sender, (mut data, recipient)) in buffer {
-            let rand_sample = rand::random::<f64>();
-            if rand_sample > self.reliability {
-                debug!("Simulated network fail.");
-                continue;
+            for peer_id in disconnected_peers.drain(..) {
+                self.peers.remove(&peer_id);
             }
-            let mut peers = self.peers.lock();
-            if let Some(peer) = peers.get_mut(&recipient) {
-                for hook in self.hook_list.lock().iter_mut() {
-                    hook.update_state(&mut data, sender, recipient);
+            for (sender, (mut data, recipient)) in buffer.drain(..) {
+                let rand_sample = rand::random::<f64>();
+                if rand_sample > self.reliability {
+                    debug!("Simulated network fail.");
+                    continue;
+                }
+                if let Some(peer) = self.peers.get_mut(&recipient) {
+                    for hook in self.hook_list.iter_mut() {
+                        hook.update_state(&mut data, sender, recipient);
+                    }
+                    peer.tx
+                        .unbounded_send((data, sender))
+                        .expect("channel should be open");
                 }
-                peer.tx
-                    .unbounded_send((data, sender))
-                    .expect("channel should be open");
             }
+            yield_now().await;
         }
-
-        Poll::Pending
     }
 }
 
-pub(crate) trait NetworkHook: Send {
-    fn update_state(&self, data: &mut NetworkData, sender: NodeIndex, recipient: NodeIndex);
-}
-
-#[derive(Clone)]
-pub(crate) struct AlertHook {
-    alerts_sent_by_connection: Arc<Mutex<HashMap<(NodeIndex, NodeIndex), usize>>>,
-}
-
-impl AlertHook {
-    pub(crate) fn new() -> Self {
-        AlertHook {
-            alerts_sent_by_connection: Arc::new(Mutex::new(HashMap::new())),
-        }
-    }
-
-    pub(crate) fn count(&self, sender: NodeIndex, recipient: NodeIndex) -> usize {
-        match self
-            .alerts_sent_by_connection
-            .lock()
-            .get(&(sender, recipient))
-        {
-            Some(count) => *count,
-            None => 0,
+async fn yield_now() {
```

Comment:
Uh, implementig our own `yield_now` is even worse than depending on tokio. I would really suggest leaving `UnreliableRouter` as-was here, with the `Future` implementation – it's unwieldy, but correct and does not need dependencies or implementations of functions that clearly should come from a library.

If you go with my suggestion of creating a separate mock subcrate, there you can simply use `tokio`, as you did previously here.